src/pybind/mgr/object_format.py

```diff
@@ -70,3 +91,131 @@ def to_yaml(obj: Any, *, many: bool = False, cls: Any = None) -> Any:
             [_obj_to_yaml(o) for o in copy], default_flow_style=False
         )
     return yaml.dump(_obj_to_yaml(copy), default_flow_style=False)
+
+
+ObjectResponseFuncType = Union[
+    Callable[..., Dict[Any, Any]],
+    Callable[..., List[Any]],
+]
+
+
+class ErrorResponseBase(Exception):
+    """An exception that can directly be converted to a mgr reponse."""
+
+    def format_response(self) -> Tuple[int, str, str]:
+        raise NotImplementedError()
+
+
+class UnknownFormat(ErrorResponseBase):
+    """Raised if the format name is unexpected.
+    This can help distinguish typos from formats that are known but
+    not implemented.
+    """
+
+    def __init__(self, format_name: str) -> None:
+        self.format_name = format_name
+
+    def format_response(self) -> Tuple[int, str, str]:
+        return -errno.EINVAL, "", f"Unknown format name: {self.format_name}"
+
+
+class UnsupportedFormat(ErrorResponseBase):
+    """Raised if the format name does not correspond to any valid
+    conversion functions.
+    """
+
+    def __init__(self, format_name: str) -> None:
+        self.format_name = format_name
+
+    def format_response(self) -> Tuple[int, str, str]:
+        return -errno.EINVAL, "", f"Unsupported format: {self.format_name}"
+
+
+class Formatter(Protocol):
+    def valid_formats(self) -> Iterable[str]:
+        ...
+
+
+class StandardFormatter:
+    """A standard formatter that can take basic objects types and convert them
+    to JSON and/or YAML.
+    """
+
+    def __init__(self, indent: Optional[int] = 2, many: bool = False) -> None:
+        self.indent = indent
+        self.many = many
+        self._valid_formats = set(
+            v.value for k, v in Format.__members__.items()
+        )
+
+    def valid_formats(self) -> Iterable[str]:
+        return self._valid_formats
+
+    def format_json(self, obj: Any) -> str:
+        return to_json(obj, indent=self.indent, many=self.many)
+
+    format_json_pretty = format_json
+
+    def format_yaml(self, obj: Any) -> str:
+        return to_yaml(obj, many=self.many)
+
+
+def formatted(
+    obj: Any,
+    format_req: Optional[str] = None,
+    formatter: Optional[Formatter] = None,
+) -> str:
+    """Return the content of the object `obj` formatted into the
+    given serialization format given by `format_req` and applied by
+    the optional formatter.
+    """
+    if format_req is None:
+        format_req = "json"  # default to json
+    if formatter is None:
+        formatter = StandardFormatter()
+    if format_req not in formatter.valid_formats():
+        raise UnknownFormat(format_req)
+    req = str(format_req).replace("-", "_")
+    ffunc = getattr(formatter, f"format_{req}", None)
+    if ffunc is None:
+        raise UnsupportedFormat(format_req)
```

Comment:
what would actually happen if somebody did not provide a `--format` arg or specified "plain" and the function had the Responder decorator on it?

Reply:
Adding code to have `format_req=None` default to something *other* than json would be doable in the future if needed FWIW.